### PR TITLE
Add rc2 ciphers to default provider

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
EVP_CIPHER has the following 2 methods
   set_asn1_parameters
   get_asn1_parameters

In legacy code most ciphers set them to either
EVP_CIPHER_set_asn1_iv,
EVP_CIPHER_get_asn1_iv

or set the flag EVP_CIPH_FLAG_DEFAULT_ASN1 (which does the same thing mostly in EVP_CIPHER_asn1_to_param())

These have custom methods for RC2 and are used by a few CMS tests.
I am not sure that they belong in the provider - they kind of need to be associated with a NID.
We could get away with it at the moment by handling them as set_params since they will be in a non-fips provider - but should we do it this way?

**Comment by levitte** [comment]
I'm not sure I have understood where the asn1_parameters routines are used or why we have them at all. CMS, you say? That's code I've barely sniffed at, so no wonder...

I wonder if we could get away with simply passing the DER blob as a suitably named parameter, and let the provider side code deal with it. This is safe to do for now, as long as we don't end up doing this in the FIPS provider.
...
I'm unsure what to do with the default handlers, though. They are used in FIPS territory, aren't they?

**Comment by slontis** [comment]
See EVP_CIPHER_param_to_asn1() & EVP_CIPHER_asn1_to_param(0 to understand how they are used..
They are mainly used to set or get an IV to/from an ASN1_TYPE..

RC2 also has a keybits field that it uses - which is why it is slightly different..
These functions are wrapped in a !FIPS_MODE

@mattcaswell any ideas welcome..

**Comment by levitte** [comment]
Oh wow...  I'm currently looking in `cms_EncryptedContent_init_bio`, and see that `X509_ALGOR` is involved somehow...  but that type is usually used for signature algorithms, right?  That doesn't usually involve *symmetric ciphers*.  From a quick reading of the CMS RFCs, I can't quite see where the algo parameters (`calg->parameter` in that function) are supposed to come from.  Mind pointing me in the right direction?
(this sounds very much like internal hackery, retrofitting a type that wasn't really meant to be used this way...  which is kind of encouraging, because that potentially means this is possible to rework in `OSSL_PARAM` terms)

**Comment by slontis** [comment]
> Mind pointing me in the right direction?

https://tools.ietf.org/html/rfc5652#section-6.1

EncryptedContentInfo ::= SEQUENCE {
        contentType ContentType,
        **contentEncryptionAlgorithm** ContentEncryptionAlgorithmIdentifier,
        encryptedContent [0] IMPLICIT EncryptedContent OPTIONAL }

**Comment by slontis** [comment]
I will look at it - Another way to do it would be to just handle this special case (which is a legacy algorithm anyway) inside the EVP layer and just call set_params to set its fields.. I dont know that handling ASN1 type in any provider is a good idea.

**Comment by levitte** [comment]
I don't have the same confidence as you that we will never see special cases like that ever again. Or well, perhaps not that kind of varying key size, as it's more fashionable to include the key size in the algo name these days, but...

**Comment by slontis** [comment]
ASN1 parsing should not be in the providers if we can help it..

**Comment by slontis** [comment]
The code in EVP_CIPHER_param_to_asn1() is doing something very interesting :)
if (cipher->prov != NULL)
   cipher = EVP_get_cipherbynid(cipher->nid);
cipher->set_asn1_parameters(c, type);

Maybe there is some way of making this nasty code work, will look into it..

**Comment by slontis** [comment]
So the way forward for this would be to make EVP_CIPHER_param_to_asn1() do ossl_param queries to the cipher in the provider.. and then do the logic in this layer based on the result.. So we have a special param just for rc2 to return its magic number..

**Comment by levitte** [comment]
I don't really see a problem with providers getting to parse a DER blob.
I *do* see the problem if that reaches the FIPS module, but that's not the case here.

**Comment by slontis** [comment]
Hmm and then we end up at some point where that may not be the case in the future..
What it is currently doing is a bit dangerous - I thought the plan would be to get rid of deprecated data in the old const EVP_CIPHER objects.

**Comment by levitte** [comment]
> I thought the plan would be to get rid of deprecated data in the old const EVP_CIPHER objects.

Uhm yes? I'm not sure what part of what we've said has you think we can't? It's true that we can't the way this is *currently* handled, but we're talking about changing that, so...

**Comment by levitte** [comment]
The commit I just pushed here adapts the provided RC2 implementation to work with #10008. 

I suggest you try it, @slontis, to see that it works for you.  I've only tested in a default build.

**Comment by levitte** [comment]
And yes, that commit does some dark OSSL_PARAM magic...  or well, uses the data in fairly standard ways, except without copying the data unnecessarily...

File: providers/default/ciphers/cipher_rc2.c [hunk 1]
```diff
@@ -51,6 +83,45 @@ static int rc2_get_ctx_params(void *vctx, OSSL_PARAM params[])
...
+        }
+        if ((type = ASN1_TYPE_new()) == NULL) {
+            ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+            return 0;
+        }
+
+        /* Is this the original IV or the running IV? */
+        num = rc2_keybits_to_magic(ctx->key_bits);
+        if (!ASN1_TYPE_set_int_octetstring(type, num,
+                                           ctx->base.iv, ctx->base.ivlen)) {
```
Comment (slontis) [inline providers/default/ciphers/cipher_rc2.c]: This looks really dangerous to me.. it was the original iv in the old code.
There is only one iv stored in the providers ctx currently (which gets changed by cipher updates for some modes)
So we either need to
1) store the oiv as well as the iv in the provider base ctx (it is currently stored in the EVP's ctx)
2) pass the EVP's original iv down as a parameter to this function.

File: providers/default/ciphers/cipher_rc2.c [hunk 2]
```diff
@@ -51,6 +83,45 @@ static int rc2_get_ctx_params(void *vctx, OSSL_PARAM params[])
...
+            ASN1_TYPE_free(type);
+            ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+            return 0;
+        }
+        /*
+         * IF the caller has a buffer, we pray to the gods they got the
```
Comment (slontis) [inline providers/default/ciphers/cipher_rc2.c]: Maybe change this comment to something a little less threatening...
The caller must call this twice - once with the buffer set to NULL, to determine the size..

File: providers/default/ciphers/cipher_rc2.c (hunk 1 above)
Comment (levitte) [reply providers/default/ciphers/cipher_rc2.c]: I think the provider side should store the original iv

Comment (slontis) [reply providers/default/ciphers/cipher_rc2.c]: Ok.. I will add a PR.

File: providers/default/ciphers/cipher_rc2.c (hunk 2 above)
Comment (levitte) [reply providers/default/ciphers/cipher_rc2.c]: Yes. If you look in #10008, that's exactly what I do. Really, this comment is a huge critique of i2d functions for not taking a buffer size parameter. But sure, the comment can be changed.

File: providers/default/ciphers/cipher_rc2.c (hunk 1 above)
Comment (levitte) [reply providers/default/ciphers/cipher_rc2.c]: Question is if that should happen in *all* implementations, or just the ones that have this custom treatment...  the latter is certainly easier!

Comment (slontis) [reply providers/default/ciphers/cipher_rc2.c]: I will do it the first way.. That is more future proof.

Comment (levitte) [reply providers/default/ciphers/cipher_rc2.c]: Awrightie then!

File: providers/default/ciphers/cipher_rc2.c [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        if (p->data_type != OSSL_PARAM_OCTET_STRING
+            || (type = d2i_ASN1_TYPE(NULL, &d, p->data_size)) == NULL
+            || ((size_t)ASN1_TYPE_get_int_octetstring(type, &num,
+                                                      ctx->base.iv,
+                                                      ctx->base.ivlen)
+                != ctx->base.ivlen)
+            || (ctx->key_bits = rc2_magic_to_keybits(num)) == 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+            ret = 0;
+        }
```
Comment (levitte) [inline providers/default/ciphers/cipher_rc2.c]: BTW, @slontis...  please review and feel free to modify how the received parameters are treated.  The pre-3.0 code does the following, I assume that something similar should happen here, just more locally...

``` C
        if (i > 0 && !EVP_CipherInit_ex(c, NULL, NULL, NULL, iv, -1))
            return -1;
        if (EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_SET_RC2_KEY_BITS, key_bits,
                                NULL) <= 0
                || EVP_CIPHER_CTX_set_key_length(c, key_bits / 8) <= 0)
            return -1;
```

Comment (levitte) [reply providers/default/ciphers/cipher_rc2.c]: (I guess it's just pure damn luck that just assigning `ctx->base.iv` and `ctx->key_bits` worked in the test case)

**Comment by slontis** [comment]
The code is dependant on #10026 AND #10008

File: providers/default/ciphers/cipher_rc2.c [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+      (void (*)(void))rc2_gettable_ctx_params },                               \
+    { OSSL_FUNC_CIPHER_SET_CTX_PARAMS,                                         \
+      (void (*)(void))rc2_set_ctx_params },                                    \
+    { OSSL_FUNC_CIPHER_SETTABLE_CTX_PARAMS,                                    \
+     (void (*)(void))rc2_settable_ctx_params },                                \
+    { 0, NULL }                                                                \
+};
+
+/* rc2128ecb_functions */
+IMPLEMENT_cipher(rc2, RC2, ecb, ECB, EVP_CIPH_VARIABLE_LENGTH, 128, 64, 0, block)
```
Comment (levitte) [inline providers/default/ciphers/cipher_rc2.c]: With the changes in #10008, you will have to add `EVP_CIPH_FLAG_CUSTOM_ASN1` to the flags.

Comment (slontis) [reply providers/default/ciphers/cipher_rc2.c]: This no longer applies as you reversed the logic..

File: crypto/err/openssl.txt [hunk 5]
```diff
@@ -2714,6 +2714,7 @@ PROV_R_TAG_NOT_NEEDED:120:tag not needed
 PROV_R_UNABLE_TO_LOAD_SHA1:143:unable to load sha1
 PROV_R_UNABLE_TO_LOAD_SHA256:147:unable to load sha256
 PROV_R_UNSUPPORTED_CEK_ALG:145:unsupported cek alg
+PROV_R_UNSUPPORTED_KEY_SIZE:150:unsupported key size
```
Comment (levitte) [inline crypto/err/openssl.txt]: According to Travis, you need to remove this line and redo `make update`

File: providers/common/include/internal/providercommonerr.h [hunk 6]
```diff
@@ -95,6 +95,7 @@ int ERR_load_PROV_strings(void);
 # define PROV_R_UNABLE_TO_LOAD_SHA1                       143
 # define PROV_R_UNABLE_TO_LOAD_SHA256                     147
 # define PROV_R_UNSUPPORTED_CEK_ALG                       145
+# define PROV_R_UNSUPPORTED_KEY_SIZE                      150
```
Comment (levitte) [inline providers/common/include/internal/providercommonerr.h]: Remove this as well before `make update`

Comment (slontis) [reply providers/common/include/internal/providercommonerr.h]: Will fix after dependent PR's are in.

**Comment by slontis** [comment]
rebased now that dependant PR's  were merged.

**Comment by slontis** [comment]
thanks. merged to master.

**Pull request closed without merging**